### PR TITLE
skip the "cancled" ddl in ddl history

### LINKED ISSUE #1228: When binlog encounters a failed DDL statement (contains if not exists / if exists), it will report an error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

When binlog encounters a failed DDL statement (contains if not exists / if exists), it will report an error.

2. What did you expect to see?

Skip these DDL

3. What did you see instead?

Report error

4. Please provide the relate downstream type and version of drainer.
(run `drainer -V` in terminal to get drainer's version)

## Reason

The reason is the failed DDL job's `SchemaVersion` is zero, which is a special case. So we can skip them.

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: close #1228

### What is changed and how it works?

as title

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### Release note

- No release note

### CODE REVIEW

File: drainer/syncer.go [hunk 1]
```diff
@@ -466,12 +466,16 @@ ForLoop:
...
-				err = errors.Trace(err)
-				break ForLoop
-			}
-			if s.cfg.SyncDDL {
+
+			if b.job.BinlogInfo.SchemaVersion == 0 {
+				log.Info("skip ddl due to the failed ddl", zap.String("sql", sql), zap.Int64("commit ts", commitTS))
+				shouldSkip = true
+			} else if s.cfg.SyncDDL {
+				schema, table, err = s.schema.getSchemaTableAndDelete(b.job.BinlogInfo.SchemaVersion)
```
Comment (lichunzhu) [inline drainer/syncer.go:474]: I think we can't skip this step when `s.cfg.SyncDDL == false`. The `version2SchemaTable` map will keep increasing and cause an oom issue. What's more, `schema, table` info will be needed when we do the following steps.

File: drainer/syncer.go [hunk 2]
```diff
@@ -466,12 +466,16 @@ ForLoop:
...
-			schema, table, err = s.schema.getSchemaTableAndDelete(b.job.BinlogInfo.SchemaVersion)
-			if err != nil {
-				err = errors.Trace(err)
-				break ForLoop
-			}
-			if s.cfg.SyncDDL {
+
+			if b.job.BinlogInfo.SchemaVersion == 0 {
+				log.Info("skip ddl due to the failed ddl", zap.String("sql", sql), zap.Int64("commit ts", commitTS))
+				shouldSkip = true
```
Comment (lichunzhu) [inline drainer/syncer.go:472]: I think just skipping this kind of jobs and then append fake binlogs is okay.
```suggestion
				appendFakeBinlogIfNeeded(nil, commitTS)
				continue
```

**Review by lichunzhu — COMMENTED** [review]
Could you please help add a unit test in `TestNewSyncer` function? Thanks a lot!

**Review by lichunzhu — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**